packages/common-common/src/cacheDecorator.ts

```diff
@@ -0,0 +1,62 @@
+import { RequestHandler, Request } from 'express';
+import { RedisNamespaces } from './types';
+import { RedisCache } from './redisCache';
+import { ServerError } from 'common-common/src/errors';
+import { defaultKeyGenerator } from './cacheKeyUtils';
+
+export class CacheDecorator {
+  private redisCache: RedisCache;
+
+  public setCache(redisCache: RedisCache) {
+    this.redisCache = redisCache;
+  }
+
+  public cache(duration: number, keyGenerator: (req: Request) => string = defaultKeyGenerator, namespace: RedisNamespaces = RedisNamespaces.Route_Response): RequestHandler {
+    return async function cacheMiddleware(req, res, next) {
+      const cacheKey = keyGenerator(req);
+      try {
+        if (!this.redisCache) {
+          next();
+          return;
+        }
+        // Try to fetch the response from Redis cache
+        const cachedResponse = await this.redisCache.getKey(namespace, cacheKey);
+        if (cachedResponse) {
+          // Response found in cache, send it
+          console.log('Response found in cache, sending it');
+          res.set('X-Cache', 'HIT');
+          res.status(200).send(JSON.parse(cachedResponse));
+          return;
+        }
+        // Response not found in cache, generate it and cache it
+        // Response not found in cache, generate it and cache it
+        const originalSend = res.send;
+        res.send = (body) => {
+          try {
+            console.log('Response not found in cache, sending it');
+            const response = originalSend.call(res,body);
+            try {
+              // const jsonBody = JSON.parse(body);
+              // if(jsonBody && jsonBody.status=='Success') {
+              if(res.statusCode == 200) {
+                this.redisCache.setKey(namespace, cacheKey, body, duration);
+              }
+            } catch (error) {
```

Comment:
Is there any way to clean up this control flow? The triple nested catch block is a bit hard to follow in terms of how the middleware will function and what the possible failure cases are.

Reply:
sorry, can try to take out this function as seperate one for readibility
- all try catch are different though
- first if some error in fetching from cache - just proceed to next middleware
- second returning erred response to user from intercepted res.send call
- third if some error writing to cache - just log and ignore

Reply:
trying this out:
https://github.com/hicommonwealth/commonwealth/pull/3492/commits/3f373b0e88de7b3590eac04bf047437f15acc47a